jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java

```diff
@@ -528,119 +530,33 @@ public Request.Processor handle(Request request) throws Exception
-        // Update headers for etags and inflation
-        HttpFields.Mutable newFields = null;
-        if (inflated || httpFields.contains(ETAG_HEADERS))
-        {
-            newFields = HttpFields.build(httpFields.size() + 1);
-            for (HttpField field : httpFields)
-            {
-                if (field.getHeader() == null)
-                {
-                    newFields.add(field);
-                    continue;
-                }
-
-                switch (field.getHeader())
-                {
-                    case IF_MATCH:
-                    case IF_NONE_MATCH:
-                    {
-                        String etags = field.getValue();
-                        String etagsNoSuffix = CompressedContentFormat.GZIP.stripSuffixes(etags);
-                        if (etagsNoSuffix.equals(etags))
-                            newFields.add(field);
-                        else
-                        {
-                            newFields.add(new HttpField(field.getHeader(), etagsNoSuffix));
-                            request.setAttribute(GZIP_HANDLER_ETAGS, etags);
-                        }
-                        break;
-                    }
-                    case CONTENT_LENGTH:
-                        newFields.add(inflated ? new HttpField("X-Content-Length", field.getValue()) : field);
-                        break;
-
-                    case CONTENT_ENCODING:
-                        if (inflated)
-                        {
-                            if (field.getValue().equalsIgnoreCase("gzip"))
-                                newFields.add(X_CE_GZIP);
-                            else if (COMMA_GZIP.matcher(field.getValue()).matches())
-                            {
-                                String v = field.getValue();
-                                v = v.substring(0, v.lastIndexOf(','));
-                                newFields.add(X_CE_GZIP);
-                                newFields.add(new HttpField(HttpHeader.CONTENT_ENCODING, v));
-                            }
-                        }
-                        else
-                        {
-                            newFields.add(field);
-                        }
-                        break;
-
-                    default:
-                        newFields.add(field);
-                }
-            }
-        }
-
-        // Don't gzip if already gzipped;
-        if (alreadyGzipped)
-        {
-            LOG.debug("{} already intercepting {}", this, request);
-            HeaderWrappingRequest wrappedRequest = new HeaderWrappingRequest(request, newFields);
-            return wrappedRequest.wrapProcessor(super.handle(wrappedRequest));
-        }
+        if (!isPathInflatable(path) && !(isPathGzipable(path) && isPathMimeTypeGzipable(path)))
+            // Bypass wrapping
+            return getHandler().handle(request);
+        return super.handle(request);
```

Comment:
Another case of "allocate the wrapper first, then decide whether to use the wrapper at all" on the hot path, for sake of possible reuse of the processor (now the logic has to be run elsewhere in case the request is different) -- I don't like it much, but let's see what benchmark says.